Draw RWD watersheds before polygon validation

kdeloach opened the pull request:
This fixes a bug where RWD shapes do not appear on the map, because the
FeatureCollection returned by RWD throws an exception during shape
validation.

Fix this by converting the watershed from a FeatureCollection to a
MultiPolygon before validation. Also decided to draw the watershed
before the validation step just in case it fails.

The rationale is that we should always err on the side of displaying the
RWD watershed, even if it fails validation. Although the shape is always
displayed now, we only navigate to the analyze view if validation
is successful.

Connects #1521

kdeloach commented:
I could adjust the RWD logic so that the watershed polygon is always drawn, before running the validation logic. Then, validate the polygon area before running analyze. 

However, I'm not sure how to accomplish the self-intersection validation with a FeatureCollection, or if the small chance of that happening is worth guarding against.

kellyi commented:
Trying to review now but having some trouble getting the rwd-data set up again.

mmcfarland commented:
@kdeloach I think that's a good compromise, I'm not sure RWD self intersections are a real problem, and if they are we could probably do something here https://github.com/WikiWatershed/rapid-watershed-delineation/blob/develop/src/api/main.py#L95 when we simplify the geometry.

@kellyi you didn't find RWD with self intersecting right, just returning multipolygons?

kellyi commented:
I don't think I saw any self-intersecting rwd shapes, but I also think they *might* have been structured as multi-polygons if they were. This PR has a few images and examples of some of the wonky shapes -- https://github.com/WikiWatershed/model-my-watershed/pull/1468 -- and it includes stuff like this where I think the multipolygon might technically self-intersect because the two adjacent parts may touch:

![image](https://cloud.githubusercontent.com/assets/4165523/18587858/022fda3e-7bf2-11e6-94ab-a33541644e2a.png)

kellyi commented:
I destroyed my worker VM and reprovisioned from scratch. Tested this and I am now able to use RWD again:

<img width="1039" alt="screen shot 2016-10-10 at 6 48 02 pm" src="https://cloud.githubusercontent.com/assets/4165523/19253584/2c29a82a-8f1a-11e6-805d-0d3c2b5981c8.png">

Here's the `response_json` for that shape:

```json
{u'watershed': {u'type': u'FeatureCollection', u'features': [{u'geometry': {u'type': u'Polygon', u'coordinates': [[[-75.39324074076636, 39.94916666663094], [-75.39240740743094, 39.94916666663094], [-75.39212962965246, 39.948981481445294], [-75.39175925928116, 39.948981481445294], [-75.391851851874, 39.94851851848119], [-75.39166666668834, 39.948425925888365], [-75.39157407409552, 39.947777777738615], [-75.39175925928116, 39.94759259255297], [-75.39305555558072, 39.94749999996015], [-75.39453703706592, 39.947037036996036], [-75.39462962965874, 39.947685185145794], [-75.3944444444731, 39.947685185145794], [-75.3944444444731, 39.94787037033144], [-75.39416666669462, 39.94805555551708], [-75.39416666669462, 39.948333333295544], [-75.3937037037305, 39.94861111107401], [-75.39351851854484, 39.949074074038116], [-75.39324074076636, 39.94916666663094]]]}, u'type': u'Feature', u'properties': {u'Avgslp': 0.089693, u'BasinLen': 0.377996, u'RR': 0.060969, u'Area': 0.038002, u'AvgOLF': 0.0, u'DrnDen': 0.0, u'BR': 23.045959, u'Strord': 1.0, u'StrLen': 0.0, u'GRIDCODE': 1}}]}, u'input_pt': {u'type': u'FeatureCollection', u'features': [{u'geometry': {u'type': u'Point', u'coordinates': [-75.39180555557758, 39.94893518514888]}, u'type': u'Feature', u'properties': {u'Lat': 39.948026, u'Dist_moved': 17, u'Lon': -75.393323, u'ID': 1}}]}}
```

I also tested the spot where I knew it would fail and got the correct error:

<img width="1030" alt="screen shot 2016-10-10 at 6 49 38 pm" src="https://cloud.githubusercontent.com/assets/4165523/19253616/59c339d6-8f1a-11e6-8e9f-6c149ac4ff3a.png">

Here's the `response_json` from the failing shapes:

```json
{u'watershed': {u'type': u'FeatureCollection', u'features': [{u'geometry': {u'type': u'MultiPolygon', u'coordinates': [[[[-75.46268518516024, 39.98120370376176], [-75.46249999997485, 39.98027777783411], [-75.46175925923333, 39.979537037091994], [-75.46222222219679, 39.978888888942635], [-75.46148148145527, 39.97805555560775], [-75.46435185182865, 39.97796296301499], [-75.46509259257017, 39.97759259264393], [-75.46546296294093, 39.97759259264393], [-75.46564814812632, 39.97814814820052], [-75.46592592590439, 39.97833333338605], [-75.46592592590439, 39.978518518571576], [-75.46611111108976, 39.978518518571576], [-75.46620370368245, 39.97907407412817], [-75.46555555553363, 39.97907407412817], [-75.46537037034824, 39.9792592593137], [-75.46425925923596, 39.979537037091994], [-75.46370370367983, 39.98009259264858], [-75.463055555531, 39.980370370426876], [-75.463055555531, 39.98064814820517], [-75.46268518516024, 39.98092592598346], [-75.46268518516024, 39.98120370376176]]], [[[-75.46259259256755, 39.98129629635452], [-75.46249999997485, 39.98129629635452], [-75.46249999997485, 39.98120370376176], [-75.46259259256755, 39.98120370376176], [-75.46259259256755, 39.98129629635452]]], [[[-75.46249999997485, 39.981388888947286], [-75.46240740738216, 39.981388888947286], [-75.46240740738216, 39.98129629635452], [-75.46249999997485, 39.98129629635452], [-75.46249999997485, 39.981388888947286]]], [[[-75.46212962960409, 39.98175925931835], [-75.4620370370114, 39.98175925931835], [-75.4620370370114, 39.98166666672558], [-75.46212962960409, 39.98166666672558], [-75.46212962960409, 39.98175925931835]]], [[[-75.46222222219679, 39.98166666672558], [-75.46212962960409, 39.98166666672558], [-75.46212962960409, 39.981574074132816], [-75.46222222219679, 39.981574074132816], [-75.46222222219679, 39.98166666672558]]], [[[-75.46240740738216, 39.98148148154005], [-75.46231481478948, 39.98148148154005], [-75.46231481478948, 39.981388888947286], [-75.46240740738216, 39.981388888947286], [-75.46240740738216, 39.98148148154005]]], [[[-75.46231481478948, 39.981574074132816], [-75.46222222219679, 39.981574074132816], [-75.46222222219679, 39.98148148154005], [-75.46231481478948, 39.98148148154005], [-75.46231481478948, 39.981574074132816]]]]}, u'type': u'Feature', u'properties': {u'Avgslp': 0.067887, u'BasinLen': 0.481181, u'RR': 0.070484, u'Area': 0.073497, u'AvgOLF': 0.739843, u'DrnDen': 0.675819, u'BR': 33.915543, u'Strord': 1.0, u'StrLen': 0.049671, u'GRIDCODE': 1}}]}, u'input_pt': {u'type': u'FeatureCollection', u'features': [{u'geometry': {u'type': u'Point', u'coordinates': [-75.4615277777516, 39.978101851904135]}, u'type': u'Feature', u'properties': {u'Lat': 39.978172, u'Dist_moved': 1, u'Lon': -75.461612, u'ID': 1}}]}}
```

kdeloach commented:
I think I found a somewhat elegant solution to this, by accident. 

I noticed that all AOI layers added to the map are mutated into MultiPolygons, so I extracted the `restructureAoI` logic into a helper, and updated the delineation logic to convert all RWD watersheds from FeatureCollection to MultiPolygon when the result comes back from the API.

However, I still decided to draw the watershed before the validation step, because I think it could still be useful to see the watershed generated from RWD even if it fails validation (we only navigate to analyze mode if validation passes).

mmcfarland commented:
That workflow sounds good, and actually anticipates some features we may utilize with the next version of RWD, namely being able to generate watersheds over very large areas just to be able to see their context.  As in, click at the mouth of the Mississippi and seeing half the country as contributing to the outlet point - even though we can't do any analysis on it.

kellyi commented:
Testing this now.

I was able to run RWD, the analysis steps, TR-55, and MapShed for this RWD shape:

![screen shot 2016-10-11 at 4 18 52 pm](https://cloud.githubusercontent.com/assets/4165523/19287199/bc09e0fe-8fce-11e6-8daa-bff591eba392.png)

->

![screen shot 2016-10-11 at 4 19 07 pm](https://cloud.githubusercontent.com/assets/4165523/19287204/c0f3d07a-8fce-11e6-8428-c741f19b98df.png)

->

![screen shot 2016-10-11 at 4 19 55 pm](https://cloud.githubusercontent.com/assets/4165523/19287208/c53f57bc-8fce-11e6-9a73-767fed6d1e0e.png)

Immediately after that I tried RWD at the place near Plumsock, PA where it should show the pop up error message. The error message was displayed:

![screen shot 2016-10-11 at 4 20 23 pm](https://cloud.githubusercontent.com/assets/4165523/19287227/dd9389a0-8fce-11e6-94ce-e79794fa3ce0.png)

When I clicked "Ok" to dismiss the box, I was brought back to the prior shape:

![screen shot 2016-10-11 at 4 20 28 pm](https://cloud.githubusercontent.com/assets/4165523/19287240/ebd8a73e-8fce-11e6-8c9b-2884f7a65f4f.png)

I was able to clear it by clicking "Reset," and it only happened when I had clicked "Delineate Watershed" while the old shape was still on screen. The shape disappeared then, but it must not have gotten fully cleared. It didn't break the app or show any errors in the console, and -- as mentioned -- I was able to clear it by clicking "Reset".

kellyi commented:
Was able to run analysis, TR-55, and MapShed for this drawn shape:

![screen shot 2016-10-11 at 4 24 50 pm](https://cloud.githubusercontent.com/assets/4165523/19287368/703a1b98-8fcf-11e6-8ec4-947993005120.png)

->

![screen shot 2016-10-11 at 4 25 08 pm](https://cloud.githubusercontent.com/assets/4165523/19287375/744133ca-8fcf-11e6-899f-cfabdfa879fc.png)

->

![screen shot 2016-10-11 at 4 25 55 pm](https://cloud.githubusercontent.com/assets/4165523/19287378/78877fac-8fcf-11e6-96d2-4524fc3a7c1d.png)

kellyi commented:
Analysis and TR-55 worked for NJ Congressional District 01, and the MapShed job timed out on the frontend (which I think is the normal behavior):

![screen shot 2016-10-11 at 4 27 30 pm](https://cloud.githubusercontent.com/assets/4165523/19287523/072af0a4-8fd0-11e6-808c-6c8e377473ee.png)

<img width="679" alt="screen shot 2016-10-11 at 4 27 58 pm" src="https://cloud.githubusercontent.com/assets/4165523/19287533/0d853ec8-8fd0-11e6-98b9-717666f34561.png">

![screen shot 2016-10-11 at 4 29 30 pm](https://cloud.githubusercontent.com/assets/4165523/19287542/1194cede-8fd0-11e6-9be8-a56d14be7216.png)

kellyi commented:
NJ-04 Congressional District worked as expected too:

![screen shot 2016-10-11 at 4 31 32 pm](https://cloud.githubusercontent.com/assets/4165523/19287591/51814aa4-8fd0-11e6-8417-35c51524b6ad.png)

![screen shot 2016-10-11 at 4 31 54 pm](https://cloud.githubusercontent.com/assets/4165523/19287597/56e16768-8fd0-11e6-880d-8249e717ef33.png)

![screen shot 2016-10-11 at 4 32 04 pm](https://cloud.githubusercontent.com/assets/4165523/19287599/5b05a8cc-8fd0-11e6-8bc0-35c0c0646696.png)

kellyi commented:
Multipolygons generated by the frontend work correctly too:

![screen shot 2016-10-11 at 4 35 26 pm](https://cloud.githubusercontent.com/assets/4165523/19287701/dbcf9f44-8fd0-11e6-86db-d8b2af148716.png)

![screen shot 2016-10-11 at 4 36 03 pm](https://cloud.githubusercontent.com/assets/4165523/19287708/df8c3052-8fd0-11e6-9cf0-3459937f0926.png)

kellyi commented:
Aside from the small quirk noted above whereby the a failed RWD shape will revert to showing the prior RWD shape, this works well! 

I think the `previousShape` might not be getting automatically cleared now after the change in the first commit, so that might now be expected behavior? I think that since technically I hadn't hit the "Reset" button to clear the prior RWD shape before trying to delineate a new one, that may actually be what the app is set up to do rather than a bug?

kellyi commented:
OK: I just checked and confirmed that an RWD failure which generates that error message will return you to non-RWD shapes when clicked, too. I selected a congressional district, then selected "Draw" and "Delineate Watershed", then picked the point where I knew it would fail. When I clicked "Ok" it sent me back to the congressional district.

I *think* this may be expected behavior now, and if so: +1. 

As I mentioned, it doesn't break the app or anything; it was just a little quirky to return to the prior shape. Basically to have that *not* happen one has to click "Reset" to clear the shape prior to clicking "Delineate Watershed".

kdeloach commented:
Originally, pressing ESC during draw mode would restore the previous layer, but that must have broken for some time (doesn't work on staging). What you're seeing is actually the expected behavior, although it may be a little out of date. 

The draw mode promise chain didn't have a validation step initially, so it made sense to restore the previous layer in the `fail` callback. However, now that the validation step is part of this chain, it might make sense to have distinct fail callbacks for drawing vs. validation.

We should address this in a separate issue.
